06-security.md

```diff
@@ -1 +1,451 @@
+
+While many of the guidelines are applicable to any REST API, there is a dedicated part for API's that are consumed via the browser (which is our typical use case). 
+
+# API Security Guidelines
+
+## Use HTTPS
+
+HTTP traffic can be easily monitored, intercepted and modified while in transit. Even if the data the API exposes does not contain sensitive data, the API keys and/or session tokens are.
+
+So, with the availability of free and easy to install [https certificates](https://letsencrypt.org/) there is no excuse for not having transport level security using HTTPS. 
+Also check the configuration with [SSLTest](https://www.ssllabs.com/ssltest/) or equivalent.
+
+## HTTP Status codes
+
+### Use HTTP 401 Unauthorized
+
+To signal that a user *is not* authenticated and needs to authenticate to access the resource.
+
+### Use HTTP 403 Forbidden 
+
+To signal that a user *is* authenticated and is *not allowed* to access the resource.
+
+## Minimize publicly available information
+
+### No brands and versions in HTTP Headers
+
+These inform potential attackers on what software is being used and make vulnerability mapping easier. 
+
+For example if the server header in the HTTP response is `Server: Apache/2.2.16 (Debian)` then the server is very likely running (the no longer supported) Debian Squeeze (6) because [that version](https://packages.debian.org/squeeze-lts/apache2) of Apache2 is packaged with it. Later versions have different Apache versions (e.g. Wheezy has 2.2.22 and Jessie 2.4.10).      
+
+### No technical details in error messages
+
+Stacktraces and vendor specific error codes tell a lot about the system and make informed guessing easier. If an error occurs these details should be logged but not presented to the user of the API. You can provide a timestamp as a reference for finding the stacktrace in the log. 
+
+### Do not repeat (erroneous) input in error messages
+
+In some conditions the browser may display the contents unescaped, potentially triggering cross site scripting. (Also this eliminates a lot of false positives when using vulnerability scanners).  
+
+### Restrict access to API documentation
+
+Unless the API is intended for public use, documentation should be available to developers only (Tools like [Swagger](http://swagger.io/) are useful when developing but should not be part of the production environment). 
+
+## Security on an API for Browser based consumption 
+
+This section will focus on the traditional session based authentication as most of our applications use a browser based user interface. 
+
+### Require Authentication on all API calls.
+
+The one notable exception being a GET request for the current authenticated user (if any). This can also be used to retrieve the XSRF token.
```

Comment:
Another possible exception is a (cheap) ping, for example for tools for operations.